Disabling the Sign task should not break publishing

It is often desirable to disable signing based on a certain condition, e.g. `onlyIf { !project.version.contains('SNAPSHOT') }`.

This works with the old `Upload`-based publishing infrastructure, but doesn't with the current publishing plugins (both Ivy and Maven).

### Example

```gradle
plugins {
    id 'java-library'
    id 'maven-publish'
    id 'signing'
}

publishing {
    publications {
        mavenJava(MavenPublication) {
            from components.java
            artifact sourcesJar
        }
    }
}

signing {
    sign publishing.publications
}

tasks.withType(Sign)*.enabled = false
```

```
$ ./gradlew publishMavenJavaPublicationToMavenLocal

> Task :publishMavenJavaPublicationToMavenLocal FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':publishMavenJavaPublicationToMavenLocal'.
> Failed to publish publication 'mavenJava' to repository 'mavenLocal'
   > Invalid publication 'mavenJava': artifact file does not exist: '/Users/marc/Desktop/signing-sample/build/libs/signing-sample-0.1.jar.asc'

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 0s
5 actionable tasks: 2 executed, 3 up-to-date
```

--- a/subprojects/ivy/src/main/java/org/gradle/api/publish/ivy/internal/publication/DefaultIvyPublication.java
+++ b/subprojects/ivy/src/main/java/org/gradle/api/publish/ivy/internal/publication/DefaultIvyPublication.java
@@ -19,6 +19,7 @@ package org.gradle.api.publish.ivy.internal.publication;
 import com.google.common.collect.Lists;
 import com.google.common.collect.Sets;
 import org.gradle.api.Action;
+import org.gradle.api.DomainObjectSet;
 import org.gradle.api.InvalidUserDataException;
 import org.gradle.api.artifacts.Dependency;
 import org.gradle.api.artifacts.DependencyArtifact;
@@ -32,6 +33,7 @@ import org.gradle.api.attributes.Usage;
 import org.gradle.api.component.ComponentWithVariants;
 import org.gradle.api.component.SoftwareComponent;
 import org.gradle.api.file.FileCollection;
+import org.gradle.api.internal.CompositeDomainObjectSet;
 import org.gradle.api.internal.FeaturePreviews;
 import org.gradle.api.internal.artifacts.DefaultModuleVersionIdentifier;
 import org.gradle.api.internal.artifacts.ivyservice.projectmodule.ProjectDependencyPublicationResolver;
@@ -56,6 +58,7 @@ import org.gradle.api.publish.ivy.internal.dependency.IvyDependencyInternal;
 import org.gradle.api.publish.ivy.internal.dependency.IvyExcludeRule;
 import org.gradle.api.publish.ivy.internal.publisher.IvyNormalizedPublication;
 import org.gradle.api.publish.ivy.internal.publisher.IvyPublicationIdentity;
+import org.gradle.api.specs.Spec;
 import org.gradle.internal.Describables;
 import org.gradle.internal.DisplayName;
 import org.gradle.internal.Factory;
@@ -94,9 +97,10 @@ public class DefaultIvyPublication implements IvyPublicationInternal {
     private final IvyModuleDescriptorSpecInternal descriptor;
     private final IvyPublicationIdentity publicationIdentity;
     private final IvyConfigurationContainer configurations;
+    private final DefaultIvyArtifactSet mainArtifacts;
+    private final PublicationArtifactSet<IvyArtifact> metadataArtifacts;
+    private final PublicationArtifactSet<IvyArtifact> derivedArtifacts;
     private final PublicationArtifactSet<IvyArtifact> publishableArtifacts;
-    private final DefaultIvyArtifactSet ivyArtifacts;
-    private final PublicationArtifactSet<IvyArtifact> additionalArtifacts;
     private final DefaultIvyDependencySet ivyDependencies;
     private final ProjectDependencyPublicationResolver projectDependencyResolver;
     private final ImmutableAttributesFactory immutableAttributesFactory;
@@ -117,9 +121,10 @@ public class DefaultIvyPublication implements IvyPublicationInternal {
         configurations = instantiator.newInstance(DefaultIvyConfigurationContainer.class, instantiator);
         this.immutableAttributesFactory = immutableAttributesFactory;
         this.featurePreviews = featurePreviews;
-        ivyArtifacts = instantiator.newInstance(DefaultIvyArtifactSet.class, name, ivyArtifactNotationParser, fileCollectionFactory);
-        additionalArtifacts = instantiator.newInstance(DefaultPublicationArtifactSet.class, IvyArtifact.class, "additional artifacts for " + name, fileCollectionFactory);
-        publishableArtifacts = new CompositePublicationArtifactSet<IvyArtifact>(IvyArtifact.class, ivyArtifacts, additionalArtifacts);
+        mainArtifacts = instantiator.newInstance(DefaultIvyArtifactSet.class, name, ivyArtifactNotationParser, fileCollectionFactory);
+        metadataArtifacts = new DefaultPublicationArtifactSet<IvyArtifact>(IvyArtifact.class, "metadata artifacts for " + name, fileCollectionFactory);
+        derivedArtifacts = new DefaultPublicationArtifactSet<IvyArtifact>(IvyArtifact.class, "derived artifacts for " + name, fileCollectionFactory);
+        publishableArtifacts = new CompositePublicationArtifactSet<IvyArtifact>(IvyArtifact.class, mainArtifacts, metadataArtifacts, derivedArtifacts);
         ivyDependencies = instantiator.newInstance(DefaultIvyDependencySet.class);
         descriptor = instantiator.newInstance(DefaultIvyModuleDescriptorSpec.class, this);
     }
@@ -151,19 +156,19 @@ public class DefaultIvyPublication implements IvyPublicationInternal {
     @Override
     public void setIvyDescriptorArtifact(IvyArtifact artifact) {
         if (this.ivyDescriptorArtifact != null) {
-            additionalArtifacts.remove(this.ivyDescriptorArtifact);
+            metadataArtifacts.remove(this.ivyDescriptorArtifact);
         }
         this.ivyDescriptorArtifact = artifact;
-        additionalArtifacts.add(artifact);
+        metadataArtifacts.add(artifact);
     }
 
     @Override
     public void setGradleModuleDescriptorArtifact(IvyArtifact artifact) {
         if (this.gradleModuleDescriptorArtifact != null) {
-            additionalArtifacts.remove(this.gradleModuleDescriptorArtifact);
+            metadataArtifacts.remove(this.gradleModuleDescriptorArtifact);
         }
         this.gradleModuleDescriptorArtifact = artifact;
-        additionalArtifacts.add(artifact);
+        metadataArtifacts.add(artifact);
     }
 
     public void descriptor(Action<? super IvyModuleDescriptorSpec> configure) {
@@ -256,22 +261,22 @@ public class DefaultIvyPublication implements IvyPublicationInternal {
     }
 
     public IvyArtifact artifact(Object source) {
-        return ivyArtifacts.artifact(source);
+        return mainArtifacts.artifact(source);
     }
 
     public IvyArtifact artifact(Object source, Action<? super IvyArtifact> config) {
-        return ivyArtifacts.artifact(source, config);
+        return mainArtifacts.artifact(source, config);
     }
 
     public void setArtifacts(Iterable<?> sources) {
-        ivyArtifacts.clear();
+        mainArtifacts.clear();
         for (Object source : sources) {
             artifact(source);
         }
     }
 
     public DefaultIvyArtifactSet getArtifacts() {
-        return ivyArtifacts;
+        return mainArtifacts;
     }
 
     public String getOrganisation() {
@@ -324,13 +329,13 @@ public class DefaultIvyPublication implements IvyPublicationInternal {
         String type = getFileExtension(file.getName());
         String extension = originalArtifact.getExtension() + "." + type;
         IvyArtifact artifact = new DefaultIvyArtifact(file, originalArtifact.getName(), extension, type, originalArtifact.getClassifier());
-        additionalArtifacts.add(artifact);
+        derivedArtifacts.add(artifact);
         return artifact;
     }
 
     @Override
     public void removeDerivedArtifact(IvyArtifact artifact) {
-        additionalArtifacts.remove(artifact);
+        derivedArtifacts.remove(artifact);
     }
 
     public IvyPublicationIdentity getIdentity() {
@@ -342,7 +347,14 @@ public class DefaultIvyPublication implements IvyPublicationInternal {
     }
 
     public IvyNormalizedPublication asNormalisedPublication() {
-        return new IvyNormalizedPublication(name, getIdentity(), getIvyDescriptorFile(), getPublishableArtifacts());
+        DomainObjectSet<IvyArtifact> existingDerivedArtifacts = this.derivedArtifacts.matching(new Spec<IvyArtifact>() {
+            @Override
+            public boolean isSatisfiedBy(IvyArtifact artifact) {
+                return artifact.getFile().exists();
+            }
+        });
+        Set<IvyArtifact> artifactsToBePublished = CompositeDomainObjectSet.create(IvyArtifact.class, mainArtifacts, metadataArtifacts, existingDerivedArtifacts);
+        return new IvyNormalizedPublication(name, getIdentity(), getIvyDescriptorFile(), artifactsToBePublished);
     }
 
     @Override

--- a/subprojects/maven/src/main/java/org/gradle/api/publish/maven/internal/publication/DefaultMavenPublication.java
+++ b/subprojects/maven/src/main/java/org/gradle/api/publish/maven/internal/publication/DefaultMavenPublication.java
@@ -21,6 +21,7 @@ import com.google.common.base.Strings;
 import com.google.common.collect.Lists;
 import com.google.common.collect.Sets;
 import org.gradle.api.Action;
+import org.gradle.api.DomainObjectSet;
 import org.gradle.api.InvalidUserDataException;
 import org.gradle.api.artifacts.DependencyArtifact;
 import org.gradle.api.artifacts.DependencyConstraint;
@@ -34,6 +35,7 @@ import org.gradle.api.attributes.Usage;
 import org.gradle.api.component.ComponentWithVariants;
 import org.gradle.api.component.SoftwareComponent;
 import org.gradle.api.file.FileCollection;
+import org.gradle.api.internal.CompositeDomainObjectSet;
 import org.gradle.api.internal.FeaturePreviews;
 import org.gradle.api.internal.artifacts.DefaultExcludeRule;
 import org.gradle.api.internal.artifacts.DefaultModuleVersionIdentifier;
@@ -103,9 +105,10 @@ public class DefaultMavenPublication implements MavenPublicationInternal {
     private final String name;
     private final MavenPomInternal pom;
     private final MavenProjectIdentity projectIdentity;
+    private final DefaultMavenArtifactSet mainArtifacts;
+    private final PublicationArtifactSet<MavenArtifact> metadataArtifacts;
+    private final PublicationArtifactSet<MavenArtifact> derivedArtifacts;
     private final PublicationArtifactSet<MavenArtifact> publishableArtifacts;
-    private final DefaultMavenArtifactSet mavenArtifacts;
-    private final PublicationArtifactSet<MavenArtifact> additionalArtifacts;
     private final Set<MavenDependencyInternal> runtimeDependencies = new LinkedHashSet<MavenDependencyInternal>();
     private final Set<MavenDependencyInternal> apiDependencies = new LinkedHashSet<MavenDependencyInternal>();
     private final Set<MavenDependency> runtimeDependencyConstraints = new LinkedHashSet<MavenDependency>();
@@ -130,9 +133,10 @@ public class DefaultMavenPublication implements MavenPublicationInternal {
         this.projectDependencyResolver = projectDependencyResolver;
         this.projectIdentity = projectIdentity;
         this.immutableAttributesFactory = immutableAttributesFactory;
-        mavenArtifacts = instantiator.newInstance(DefaultMavenArtifactSet.class, name, mavenArtifactParser, fileCollectionFactory);
-        additionalArtifacts = new DefaultPublicationArtifactSet<MavenArtifact>(MavenArtifact.class, "additional artifacts for " + name, fileCollectionFactory);
-        publishableArtifacts = new CompositePublicationArtifactSet<MavenArtifact>(MavenArtifact.class, mavenArtifacts, additionalArtifacts);
+        mainArtifacts = instantiator.newInstance(DefaultMavenArtifactSet.class, name, mavenArtifactParser, fileCollectionFactory);
+        metadataArtifacts = new DefaultPublicationArtifactSet<MavenArtifact>(MavenArtifact.class, "metadata artifacts for " + name, fileCollectionFactory);
+        derivedArtifacts = new DefaultPublicationArtifactSet<MavenArtifact>(MavenArtifact.class, "derived artifacts for " + name, fileCollectionFactory);
+        publishableArtifacts = new CompositePublicationArtifactSet<MavenArtifact>(MavenArtifact.class, mainArtifacts, metadataArtifacts, derivedArtifacts);
         pom = instantiator.newInstance(DefaultMavenPom.class, this);
         this.featurePreviews = featurePreviews;
     }
@@ -164,19 +168,19 @@ public class DefaultMavenPublication implements MavenPublicationInternal {
     @Override
     public void setPomArtifact(MavenArtifact artifact) {
         if (this.pomArtifact != null) {
-            additionalArtifacts.remove(this.pomArtifact);
+            metadataArtifacts.remove(this.pomArtifact);
         }
         this.pomArtifact = artifact;
-        additionalArtifacts.add(artifact);
+        metadataArtifacts.add(artifact);
     }
 
     @Override
     public void setGradleModuleMetadataArtifact(MavenArtifact artifact) {
         if (this.moduleMetadataArtifact != null) {
-            additionalArtifacts.remove(this.moduleMetadataArtifact);
+            metadataArtifacts.remove(this.moduleMetadataArtifact);
         }
         this.moduleMetadataArtifact = artifact;
-        additionalArtifacts.add(artifact);
+        metadataArtifacts.add(artifact);
     }
 
     public void pom(Action<? super MavenPom> configure) {
@@ -279,21 +283,21 @@ public class DefaultMavenPublication implements MavenPublicationInternal {
     }
 
     public MavenArtifact artifact(Object source) {
-        return mavenArtifacts.artifact(source);
+        return mainArtifacts.artifact(source);
     }
 
     public MavenArtifact artifact(Object source, Action<? super MavenArtifact> config) {
-        return mavenArtifacts.artifact(source, config);
+        return mainArtifacts.artifact(source, config);
     }
 
     public MavenArtifactSet getArtifacts() {
         populateFromComponent();
-        return mavenArtifacts;
+        return mainArtifacts;
     }
 
     public void setArtifacts(Iterable<?> sources) {
         artifactsOverridden = true;
-        mavenArtifacts.clear();
+        mainArtifacts.clear();
         for (Object source : sources) {
             artifact(source);
         }
@@ -347,13 +351,13 @@ public class DefaultMavenPublication implements MavenPublicationInternal {
     @Override
     public MavenArtifact addDerivedArtifact(MavenArtifact originalArtifact, Factory<File> file) {
         MavenArtifact artifact = new DerivedMavenArtifact(originalArtifact, file);
-        additionalArtifacts.add(artifact);
+        derivedArtifacts.add(artifact);
         return artifact;
     }
 
     @Override
     public void removeDerivedArtifact(MavenArtifact artifact) {
-        additionalArtifacts.remove(artifact);
+        derivedArtifacts.remove(artifact);
     }
 
     public MavenProjectIdentity getMavenProjectIdentity() {
@@ -384,7 +388,14 @@ public class DefaultMavenPublication implements MavenPublicationInternal {
 
     public MavenNormalizedPublication asNormalisedPublication() {
         populateFromComponent();
-        return new MavenNormalizedPublication(name, pom.getPackaging(), getPomArtifact(), projectIdentity, getPublishableArtifacts(), determineMainArtifact());
+        DomainObjectSet<MavenArtifact> existingDerivedArtifacts = this.derivedArtifacts.matching(new Spec<MavenArtifact>() {
+            @Override
+            public boolean isSatisfiedBy(MavenArtifact artifact) {
+                return artifact.getFile().exists();
+            }
+        });
+        Set<MavenArtifact> artifactsToBePublished = CompositeDomainObjectSet.create(MavenArtifact.class, mainArtifacts, metadataArtifacts, existingDerivedArtifacts);
+        return new MavenNormalizedPublication(name, pom.getPackaging(), getPomArtifact(), projectIdentity, artifactsToBePublished, determineMainArtifact());
     }
 
     private MavenArtifact getPomArtifact() {
@@ -423,7 +434,7 @@ public class DefaultMavenPublication implements MavenPublicationInternal {
 
     private Set<MavenArtifact> getUnclassifiedArtifactsWithExtension() {
         populateFromComponent();
-        return CollectionUtils.filter(mavenArtifacts, new Spec<MavenArtifact>() {
+        return CollectionUtils.filter(mainArtifacts, new Spec<MavenArtifact>() {
             public boolean isSatisfiedBy(MavenArtifact mavenArtifact) {
                 return hasNoClassifier(mavenArtifact) && hasExtension(mavenArtifact);
             }
@@ -505,7 +516,7 @@ public class DefaultMavenPublication implements MavenPublicationInternal {
      */
     private void checkThatArtifactIsPublishedUnmodified(PublishArtifact source) {
         populateFromComponent();
-        for (MavenArtifact mavenArtifact : mavenArtifacts) {
+        for (MavenArtifact mavenArtifact : mainArtifacts) {
             if (source.getFile().equals(mavenArtifact.getFile())
                 && source.getExtension().equals(mavenArtifact.getExtension())
                 && Strings.nullToEmpty(source.getClassifier()).equals(Strings.nullToEmpty(mavenArtifact.getClassifier()))) {

--- a/subprojects/publish/src/main/java/org/gradle/api/publish/internal/PublicationInternal.java
+++ b/subprojects/publish/src/main/java/org/gradle/api/publish/internal/PublicationInternal.java
@@ -42,6 +42,19 @@ public interface PublicationInternal<T extends PublicationArtifact> extends Publ
 
     void whenPublishableArtifactRemoved(Action<? super T> action);
 
+    /**
+     * Add a derived artifact for the supplied original artifact.
+     *
+     * <p>Derived artifacts are not mandatory, i.e. when the supplied file does not exist when this
+     * publication is about to be published, they will simply be omitted from the file transfer.
+     *
+     * <p>Currently, the only known use case for derived artifacts is adding signature files
+     * created by the signing plugin.
+     *
+     * @param originalArtifact The original artifact to create a derived artifact for.
+     * @param file The file to be used for publishing the derived artifact.
+     * @return The newly created derived artifact.
+     */
     T addDerivedArtifact(T originalArtifact, Factory<File> file);
 
     void removeDerivedArtifact(T artifact);

--- a/subprojects/signing/src/integTest/groovy/org/gradle/plugins/signing/SigningPublicationsIntegrationSpec.groovy
+++ b/subprojects/signing/src/integTest/groovy/org/gradle/plugins/signing/SigningPublicationsIntegrationSpec.groovy
@@ -17,6 +17,7 @@
 package org.gradle.plugins.signing
 
 import org.gradle.test.fixtures.file.TestFile
+import spock.lang.Issue
 
 class SigningPublicationsIntegrationSpec extends SigningIntegrationSpec {
 
@@ -463,5 +464,65 @@ class SigningPublicationsIntegrationSpec extends SigningIntegrationSpec {
         then:
         fails "signIvyPublication"
         failureDescriptionContains "Task 'signIvyPublication' not found"
+    }
+
+    @Issue("https://github.com/gradle/gradle/issues/5099")
+    def "disabling sign tasks skips uploading signature artifacts but does not break publishing"() {
+        given:
+        buildFile << """
+            apply plugin: 'ivy-publish'
+            apply plugin: 'maven-publish'
+            ${keyInfo.addAsPropertiesScript()}
+
+            publishing {
+                publications {
+                    ivy(IvyPublication) {
+                        from components.java
+                        module '$artifactId'
+                    }
+                    maven(MavenPublication) {
+                        from components.java
+                        artifactId '$artifactId'
+                    }
+                }
+                repositories {
+                    maven {
+                        url "file://\$buildDir/m2Repo/"
+                    }
+                    ivy {
+                        url "file://\$buildDir/ivyRepo/"
+                        layout "pattern"
+                        artifactPattern "\$buildDir/ivyRepo/[artifact]-[revision](-[classifier])(.[ext])"
+                        ivyPattern "\$buildDir/ivyRepo/[artifact]-[revision](-[classifier])(.[ext])"
+                    }
+                }
+            }
+
+            signing {
+                ${signingConfiguration()}
+                sign publishing.publications
+            }
+
+            tasks.withType(Sign)*.enabled = false
+        """
+
+        when:
+        succeeds "publish"
+
+        then:
+        ":signIvyPublication" in skippedTasks
+        ":signMavenPublication" in skippedTasks
+        ":publishIvyPublicationToIvyRepository" in nonSkippedTasks
+        ":publishMavenPublicationToMavenRepository" in nonSkippedTasks
+
+        and:
+        pom().assertExists()
+        pomSignature().assertDoesNotExist()
+        m2RepoFile(jarFileName).assertExists()
+        m2RepoFile("${jarFileName}.asc").assertDoesNotExist()
+        ivyRepoFile(jarFileName).assertExists()
+        ivyRepoFile("${jarFileName}.asc").assertDoesNotExist()
+        ivyRepoFile("ivy-${version}.xml").assertExists()
+        ivyRepoFile("ivy-${version}.xml.asc").assertDoesNotExist()
     }
 }

